core/keyboardshortcut_registry.js

```diff
@@ -0,0 +1,351 @@
+ *     Blockly.KeyboardShortcutRegistry.registry.createSerializedKey;
+ * @param {string} shortcutName The name of the shortcut to execute when the
+ *     given keycode is pressed.
+ * @param {boolean=} opt_allowCollision True to prevent an error when adding a
+ *     shortcut to a key that is already mapped to a shortcut.
+ * @throws {Error} if the given key code is already mapped to a shortcut.
+ * @public
+ */
+Blockly.KeyboardShortcutRegistry.prototype.addKeyMapping = function(
+    keyCode, shortcutName, opt_allowCollision) {
+  var shortcutNames = this.keyMap_[keyCode];
+  if (shortcutNames && !opt_allowCollision) {
+    throw new Error(
+        'Shortcut with name "' + shortcutName + '" collides with shortcuts ' +
+        shortcutNames.toString());
+  } else if (shortcutNames && opt_allowCollision) {
+    shortcutNames.unshift(shortcutName);
+  } else {
+    this.keyMap_[keyCode] = [shortcutName];
+  }
+};
+
+/**
+ * Adds a mapping between a keycode and a keyboard shortcut.
+ * @param {string} keyCode The key code for the keyboard shortcut. If
+ *     registering a key code with a modifier (ex: ctrl+c) use
+ *     Blockly.KeyboardShortcutRegistry.registry.createSerializedKey;
+ * @param {string} shortcutName The name of the shortcut to execute when the
+ *     given keycode is pressed.
+ * @param {boolean=} opt_quiet True to not console warn when there is no
+ *     shortcut to remove.
+ * @return {boolean} True if a key mapping was removed, false otherwise.
+ * @public
+ */
+Blockly.KeyboardShortcutRegistry.prototype.removeKeyMapping = function(
+    keyCode, shortcutName, opt_quiet) {
+  var shortcutNames = this.keyMap_[keyCode];
+
+  if (!shortcutNames && !opt_quiet) {
+    console.warn(
+        'No keyboard shortcut with name "' + shortcutName +
+        '" registered with key code "' + keyCode + '"');
+    return false;
+  }
+
+  var shortcutIdx = shortcutNames.indexOf(shortcutName);
+  if (shortcutIdx > -1) {
+    shortcutNames.splice(shortcutIdx, 1);
+    if (shortcutNames.length == 0) {
+      delete this.keyMap_[keyCode];
+      return true;
+    }
+  } else if (!opt_quiet) {
+    console.warn(
+        'No keyboard shortcut with name "' + shortcutName +
+        '" registered with key code "' + keyCode + '"');
+    return false;
+  }
+};
+
+/**
+ * Sets the key map. Setting the key map will override any default key mappings.
+ * @param {!Object<string, !Array<string>>} keyMap The object with key code to
+ *     shortcut names.
+ * @public
+ */
+Blockly.KeyboardShortcutRegistry.prototype.setKeyMap = function(keyMap) {
+  this.keyMap_ = keyMap;
+};
+
+/**
+ * Gets the current key map.
+ * @return {!Object<string,!Array<!Blockly.KeyboardShortcutRegistry.KeyboardShortcut>>}
+ *     The object holding key codes to
+ *     Blockly.KeyboardShortcutRegistry.KeyboardShortcut.
+ * @public
+ */
+Blockly.KeyboardShortcutRegistry.prototype.getKeyMap = function() {
+  var keyMap = {};
```

Comment:
Curious why we're doing a copy?
Also just an FYI, mixin doesn't do a deep copy.

Reply:
From what I can remember this was to prevent people from accidentally changing the key map. They would need to use `setKeyMap` to update it. I don't have super strong feelings either way, but it seems nice to have. 

I think I can accomplish this by calling `Blockly.utils.object.deepMerge` instead of `Blockly.utils.object.mixin`.